### DISCUSSION TITLE
Where can I find the locations of the mesh vertices (knots)?

### DISCUSSION THREAD

**Discussion by ericward-noaa** [sdmTMB/sdmTMB#230] [Q&A] [answered]
Question from workshop: I would like to make predictions at the locations of the vertices on my mesh. How can I do that?

**Comment by ericward-noaa** [comment, marked as answer]
Build a mesh and fit the initial model, 

`
mesh <- make_mesh(pcod_2011, c("X", "Y"), cutoff = 20)
`

`fit <- sdmTMB(
  density ~ s(depth),
  data = pcod_2011, mesh = mesh,
  family = tweedie(link = "log")
)`

Next, build a prediction dataframe based on the locations of the vertices

`
df <- data.frame(X = mesh$mesh$loc[,1], Y = mesh$mesh$loc[,2], 
depth = mean(pcod_2011$depth))
`

And finally predict to this new dataframe,
`
pred <- predict(fit, newdata = df, se_fit = TRUE)
`

**Comment by seananderson** [comment]
I'll just note that some elements with their standard errors are already computed at the vertices/knots when the model is fit:

``` r
library(sdmTMB)
mesh <- make_mesh(pcod_2011, c("X", "Y"), cutoff = 20)
fit <- sdmTMB(
  density ~ 1, 
  data = pcod_2011, 
  mesh = mesh, 
  family = tweedie(link = "log"),
  time = "year")

est <- as.list(fit$sd_report, "Est")
se <- as.list(fit$sd_report, "Std. Error")

# spatial random field values:
head(est$omega_s)
#>             [,1]
#> [1,]  0.01722212
#> [2,] -2.02114806
#> [3,] -0.64270859
#> [4,] -1.94454560
#> [5,] -1.55586713
#> [6,] -1.02392707

# spatiotemporal random field values:
head(est$epsilon_st) # each column is a slice of time
#> , , 1
#> 
#>            [,1]         [,2]        [,3]         [,4]
#> [1,]  0.9831283 -0.004693625 -0.08555185 -0.886773307
#> [2,] -0.6931135 -0.004098653 -0.02702732  0.007235493
#> [3,]  0.2373814  0.840202893 -0.46125731 -0.844328402
#> [4,] -0.2087537 -0.145041771 -0.03636193 -0.299671830
#> [5,] -0.8138137  0.113691324  0.45828899 -0.310111808
#> [6,] -0.2087837 -0.142492183  0.18905023 -0.201013368

# equivalent standard errors:

head(se$omega_s)
#>           [,1]
#> [1,] 0.9723077
#> [2,] 1.1811803
#> [3,] 0.7782864
#> [4,] 1.3329119
#> [5,] 0.9364035
#> [6,] 1.5553873
head(se$epsilon_st)
#> , , 1
#> 
#>           [,1]      [,2]      [,3]      [,4]
#> [1,] 0.9196379 0.9533639 1.0791278 1.0223345
#> [2,] 0.9611217 0.9333101 0.9482070 0.9260709
#> [3,] 0.7904982 0.7716381 0.8084578 0.8057576
#> [4,] 0.8342544 0.8496581 0.8475800 0.8502862
#> [5,] 1.0590750 0.9396540 0.9057445 0.9800289
#> [6,] 1.0175840 1.0235955 1.0369448 1.0430628
```

<sup>Created on 2023-05-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ericward-noaa** [comment]
As a follow up -- is there a way to separate out the mesh border areas from the mesh interior area?

**Comment by ericward-noaa** [comment]
With a mesh created by INLA, there are some functions that let you extract the boundary / interior, e.g. 

`mesh <- make_mesh(pcod, c("X", "Y"), cutoff = 30, type = "cutoff")` 
`boundary <- INLA::inla.mesh.boundary(mesh$mesh)`
`interior <- INLA::inla.mesh.interior(mesh$mesh)`

In this case, there's only the boundary extracted. There's 51 vertices total,

`
 mesh$mesh$n
`

but 
`boundary[[1]]$loc`
gives the locations of the boundary points (n = 23)

In the mesh objects, the boundary points are always appended to the locations after the non-boundary points. So we can extract the non-boundary by removing the last 23 observations, e.g.

`end_point <- mesh$mesh$n - nrow(boundary[[1]]$loc)`
`mesh$mesh$loc[1:end_point,]`
